designs/2024-per-rule-autofix-configuration/README.md

```diff
@@ -0,0 +1,160 @@
+- Repo: <https://github.com/eslint/eslint>
+- Start Date: 2024-10-22
+- RFC PR:
+- Authors: [Samuel Therrien](https://github.com/Samuel-Therrien-Beslogic) (aka [@Avasam](https://github.com/Avasam))
+
+# Per-rule autofix configuration
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+This feature aims to make it possible to control autofixes through shareable configuration on a per-rule basis.
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+Some rules provide autofixing, which is great, but can sometimes be broken or otherwise simply unwanted for various reasons.
+Unsafe autofixes should be suggestions, and broken fixes should be reported, *but* ESLint is a large ecosystem where some very useful plugins are not always actively maintained. Even then, wanting to disable an autofix for project-specific or personal reasons could still happen.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+Similar to how Ruff (<https://docs.astral.sh/ruff/settings/#lint_unfixable>) does it, a top-level key to specify which rules to not autofix would be in my opinion the least disruptive and forward/backwards compatible. It should be overridable (in the `overrides` section), and picked up when extending a configuration.
+
+Concretely, it could look like this:
+
+```js
+export default [
+  {
+    autofixes: {
+      // We don't want this to autofix, as a rule suddenly not failing should require human attention
+      "@eslint-community/eslint-comments/no-unused-disable": false,
+    },
+    rules: {
+      '@eslint-community/eslint-comments/no-unused-disable': 'error',
+    }
+    overrides: [
+      files: ["*.spec.js"],
+      autofixes: {
+        // Let's pretend we want this to be autofixed in tests, for the sake of the RFC
+        "@eslint-community/eslint-comments/no-unused-disable": true,
+      },
+    ]
+  }
+]
+```
+
+The fix should still exist as a suggestion. Only autofixing (when running `eslint --fix` or editor action on save) should be disabled.
```

Comment:
I'm assuming you mean in all situations when autofixes are applied? That would include using the `ESLint.outputFixes()` or `SourceCodeFixer` APIs.

Reply:
I am not familiar with ESLint's API or implementation, how would you rephrase this? I wrote this purely from a user's PoV because, well, that's what I am ^^

Reply:
Technically, `Linter` (or `report-translator`?) could just create a new object in the `suggestions` array and move the `fix` into it. We'd just have to figure out what to set as `desc` (the text that appears in IDEs as an action to apply the suggestion).

https://eslint.org/docs/latest/integrate/nodejs-api#-lintmessage-type

Reply:
That is a good point :-)

Reply:
"Fix this autofix/prefer-template problem" is a text VS Code generates by itself when there's `LintMessge#fix` property. On click, VS Code applies the code change described in `LintMessage#fix`.

For `LintMessage#suggestions`, VS Code shows `suggestion.desc` as the text. On click, VS Code applies the code change described in `suggestion.fix`.

So, IDEs use both `LintMessge#fix` and `LintMessage#suggestions` to generate editor suggestions. Rules don't provide the same fix in both `LintMessage#fix` and `LintMessage#suggestions`. Furthermore, it probably never makes sense for a rule to provide `LintMessage#suggestions` when it provides `LintMessage#fix`.

Reply:
I think it makes sense to provide both sometimes - an autofix can be partial, and suggestions can go farther.

Reply:
Maybe just `desc: 'Apply autofix'`?

> or maybe allowing an empty `desc` to fallback to a default "fix this" message should be a requirement for this RFC.

Suggestions should generally provide a descriptive message, so I think it's better to not implement a fallback.